hardware_drivers/navigator_kill_board/nodes/kill_board_driver.py

```diff
@@ -81,12 +97,51 @@ def run(self):
     def update_ros(self):
         self.update_hw_kill()
         self.publish_diagnostics()
+        if self.ctrl_msg_received is True:
+            self.publish_joy()
+            self.ctrl_msg_received = False
 
     def handle_byte(self, msg):
         '''
         React to a byte recieved from the board. This could by an async update of a kill status or
         a known response to a recent request
         '''
+        # If the controller message start byte is received, next 8 bytes are the controller data
+        if msg == constants['CONTROLLER']:
+            self.ctrl_msg_count = 8
+            self.ctrl_msg_timeout = rospy.Time.now()
+            return
+        # If receiving the controller message, record the byte as stick/button data
+        if (self.ctrl_msg_count > 0) and (self.ctrl_msg_count <= 8):
+            # If 1 second has passed since the message began, timeout and report warning
+            if (rospy.Time.now() - self.ctrl_msg_timeout) >= rospy.Duration(1):
+                self.ctrl_msg_received = False
+                self.ctrl_msg_count = 0
+                rospy.logwarn('Timeout receiving controller message. Please disconnect controller.')
+            if self.ctrl_msg_count > 2:  # The first 6 bytes in the message are stick data bytes
+                if (self.ctrl_msg_count % 2) == 0:  # Even number byte: first byte in data word
+                    self.sticks_temp = (int(msg.encode("hex"), 16) << 8)
+                else:  # Odd number byte: combine two bytes into a stick's data word
+                    self.sticks_temp += int(msg.encode("hex"), 16)
+                    if (self.ctrl_msg_count > 6):
+                        self.sticks['UD'] = self.sticks_temp
+                    elif (self.ctrl_msg_count > 4):
+                        self.sticks['LR'] = self.sticks_temp
+                    else:
+                        self.sticks['TQ'] = self.sticks_temp
+                    self.sticks_temp = 0x0000
+            else:  # The last 2 bytes are button data bytes
```

Comment:
You mean first 2?

Reply:
As described above, these are the last 2. This logically works out, but I see how this can cause confusion. I'll rewrite it to be more readable, with the count increasing as it reads bytes.